src/qibo/models/circuit.py

```diff
@@ -300,32 +300,33 @@ def wire_names(self, wire_names: Union[list, dict]):
                     f"but is {len(wire_names)}.",
                 )
 
-            if any([not isinstance(name, str) for name in wire_names]):
-                raise_error(ValueError, "all wire names must be type ``str``.")
-
-            self._wire_names = wire_names
+            self._wire_names = wire_names.copy()
         elif isinstance(wire_names, dict):
             if len(wire_names.keys()) > self.nqubits:
                 raise_error(
                     ValueError,
                     "number of elements in the ``wire_names`` dictionary "
                     + "cannot be bigger than ``nqubits``.",
                 )
-
-            if any([not isinstance(name, str) for name in wire_names.keys()]) or any(
-                [not isinstance(name, str) for name in wire_names.values()]
-            ):
+            if not all(isinstance(k, int) for k in wire_names.values()):
                 raise_error(
                     ValueError,
-                    "all keys and values in the ``wire_names`` dictionary must be type ``str``.",
+                    "all values of the ``wire_names`` dictionary must be integers.",
+                )
+            if sorted(wire_names.values()) != list(range(self.nqubits)):
+                raise_error(
+                    ValueError,
+                    "all values of the ``wire_names`` dictionary must be unique integers from 0 to ``nqubits``.",
                 )
 
             self._wire_names = [
-                wire_names.get(f"q{i}", f"q{i}") for i in range(self.nqubits)
+                k for k, _ in sorted(wire_names.items(), key=lambda x: x[1])
             ]
         else:
             self._wire_names = [f"q{i}" for i in range(self.nqubits)]
 
+        self.init_kwargs["wire_names"] = self._wire_names
```

Comment:
The default could be moved to the getter (the `@property` decorated one), and we could drop the `"q{}"` mangling.

Reply:
When we draw the circuit, it will draw like this when `wire_names` is not set `(circuit._wire_names is None)`.

```py
"0: ─H─U1─U1─U1─U1───────────────────────────x───\n"
"1: ───o──|──|──|──H─U1─U1─U1────────────────|─x─\n"
"2: ──────o──|──|────o──|──|──H─U1─U1────────|─|─\n"
"3: ─────────o──|───────o──|────o──|──H─U1───|─x─\n"
"4: ────────────o──────────o───────o────o──H─x───"
```

Should add an additional q in that cases `(circuit._wire_names is None)`?

```py
"q0: ─H─U1─U1─U1─U1───────────────────────────x───\n"
"q1: ───o──|──|──|──H─U1─U1─U1────────────────|─x─\n"
"q2: ──────o──|──|────o──|──|──H─U1─U1────────|─|─\n"
"q3: ─────────o──|───────o──|────o──|──H─U1───|─x─\n"
"q4: ────────────o──────────o───────o────o──H─x───"
```

Reply:
I'm not sure. To be fair, to me, it's fine both ways: one will be more familiar (with `"q"`s), the other one more accurate (reflecting the actual labels used).

It is mostly aesthetics, but let's ask for @stavros11 opinion.

Reply:
> It is mostly aesthetics, but let's ask for @stavros11 opinion.

We discussed this yesterday with @csookim and I believe we will go without `"q"`s for simplicity. Looking at some changes, I think this has also been implemented.